docs/proposals/20220414-runtime-hooks.md

```diff
@@ -0,0 +1,430 @@
+
+####  Before Cluster Delete
+
+This hook is called after the Cluster has been deleted by the user, and immediately before objects existing in the Cluster are going to be deleted. Runtime Extension implementers can use this hook to execute cleanup tasks for the add-ons and block deletion of the Cluster and descendant objects until everything is ready. 
+
+##### Example Request:
+
+```yaml
+apiVersion: hooks.runtime.cluster.x-k8s.io/v1alpha1
+kind: BeforeClusterDeleteRequest
+cluster:
+  apiVersion: cluster.x-k8s.io/v1beta1
+  kind: Cluster
+  metadata:
+   name: test-cluster
+   namespace: test-ns
+  spec:
+   ...
+  status:
+   ...
+```
+
+##### Example Response:
+
+```yaml
+apiVersion: hooks.runtime.cluster.x-k8s.io/v1alpha1
+kind: BeforeClusterDeleteResponse
+status: Success
+message: "error message if status == Failure"
+retryAfterSeconds: 10
+```
+
+For additional details, refer to the [Draft OpenAPI spec](https://editor.swagger.io/?url=https://gist.githubusercontent.com/ykakarap/6a58d976df90928cdb31768125d73613/raw/8ce6175060039c2fb7b3ea9a091d712228f56852/lifecycle-hooks-openapi-2.yaml).
+
+
+###  Runtime Extensions developer guide
+
+All guidelines defined in the [Runtime SDK](https://github.com/kubernetes-sigs/cluster-api/blob/b48a6ed07ac2bd353f99000270510369f4baa1a5/docs/proposals/20220221-runtime-SDK.md) apply to the implementation of Runtime Extensions of the hooks defined in this proposal. 
+
+TL;DR; Runtime Extensions are components that should be designed, written and deployed with great caution given that they can affect the proper functioning of the Cluster API runtime. A poorly implemented Runtime Extension could potentially block lifecycle transitions from happening.
+
+Following recommendations are especially relevant:
+
+* [Blocking and non Blocking](https://github.com/kubernetes-sigs/cluster-api/blob/b48a6ed07ac2bd353f99000270510369f4baa1a5/docs/proposals/20220221-runtime-SDK.md#blocking-hooks)
+* [Error management](https://github.com/kubernetes-sigs/cluster-api/blob/b48a6ed07ac2bd353f99000270510369f4baa1a5/docs/proposals/20220221-runtime-SDK.md#error-management)
+* [Avoid dependencies](https://github.com/kubernetes-sigs/cluster-api/blob/b48a6ed07ac2bd353f99000270510369f4baa1a5/docs/proposals/20220221-runtime-SDK.md#avoid-dependencies)
+
+
+###  Security Model
+
+For the general Runtime Extension security model please refer to the [developer guide in the Runtime SDK proposal](https://github.com/kubernetes-sigs/cluster-api/blob/75b39db545ae439f4f6203b5e07496d3b0a6aa75/docs/proposals/20220221-runtime-SDK.md#security-model).
+
+###  Risks and Mitigations
+
+####  Runtime Extension blocks Cluster lifecycle indefinitely
+
+Cluster lifecycle can be blocked indefinitely when a Runtime Extension either blocks or fails indefinitely. Mitigation:
+
+* A Runtime Extension should be extensively unit and e2e tested to ensure it behaves as expected.
```

Comment:
I'm not convinced adding test to ensure the risk doesn't happen is a real mitigation... Maybe we should instead consider what happens if it does fail. How do we alert the user? Is there a timeout option available to proceed after a certain time? How can an operator unblock the cluster lifecycle via manual intervention?

Reply:
> How do we alert the user?

I think that is definitely something we can & should do (and describe here a bit). Our current plan was to surface this at least via conditions.  I assume this would then also propagate to some metrics exposed by cluster-api-state-metrics later on.

> Is there a timeout option available to proceed after a certain time?

Currently not.

> How can an operator unblock the cluster lifecycle via manual intervention?

Once a user saw that the cluster lifecycle is blocked via conditions / metrics (/alerts) he can either remove the corresponding ExtensionConfig or modify it's namespaceSelector so it doesn't match the blocked cluster anymore. Good point to mention that here.

Reply:
+1 to make it explicit that we are currently thinking at condition metrics, but in the future we can explore things like circuit breakers or timeout as suggested by @CecileRobertMichon